src/filter/http_context.rs

```diff
@@ -1,159 +1,259 @@
-        actions: &[RLA_action_specifier],
-    ) -> Result<RateLimitRequest, UtilsErr> {
+    fn process_rate_limit_policy(&self, rlp: &RateLimitPolicy) -> Action {
+        let descriptors = self.build_descriptors(rlp);
+        if descriptors.len() == 0 {
+            debug!("[context_id: {}] empty descriptors", self.context_id);
+            return Action::Continue;
+        }
+
         let mut rl_req = RateLimitRequest::new();
+        rl_req.set_domain(rlp.rate_limit_domain.clone());
+        rl_req.set_hits_addend(1);
+        rl_req.set_descriptors(descriptors);
 
-        rl_req.set_domain(domain.into());
+        let rl_req_serialized = Message::write_to_bytes(&rl_req).unwrap(); // TODO(rahulanand16nov): Error Handling
 
-        rl_req.set_hits_addend(1);
+        match self.dispatch_grpc_call(
+            rlp.upstream_cluster.as_str(),
+            RATELIMIT_SERVICE_NAME,
+            RATELIMIT_METHOD_NAME,
+            Vec::new(),
+            Some(&rl_req_serialized),
+            Duration::from_secs(5),
+        ) {
+            Ok(call_id) => info!("Initiated gRPC call (id# {}) to Limitador", call_id),
+            Err(e) => {
+                warn!("gRPC call to Limitador failed! {:?}", e);
+                request_process_failure(self.config.failure_mode_deny);
+            }
+        }
+        Action::Pause
+    }
 
-        rl_req
-            .mut_descriptors()
-            .push(descriptor_from_actions(self, actions)?);
+    fn build_descriptors(
+        &self,
+        rlp: &RateLimitPolicy,
+    ) -> ::protobuf::RepeatedField<RateLimitDescriptor> {
+        //::protobuf::RepeatedField::default()
+        rlp.gateway_actions
+            .iter()
+            .filter(|ga| self.filter_configurations_by_rules(&ga.rules))
+            // filter None and return configurations
+            .flat_map(|ga| &ga.configurations)
+            // flatten the vec<vec<Configurations> to vec<Configuration>
+            .flatten()
+            // filter None and return actions.
+            // All actions cannot be flatten! each vec of actions defines one potential descriptor
+            .flat_map(|configuration| &configuration.actions)
+            // filter None and return descriptors
+            .flat_map(|actions| self.build_single_descriptor(actions))
+            .collect()
+    }
 
-        Ok(rl_req)
+    fn filter_configurations_by_rules(&self, rules_opt: &Option<Vec<Rule>>) -> bool {
+        match rules_opt {
+            // no rules is equivalent to matching all the requests.
+            None => true,
+            Some(rules) => rules.iter().any(|rule| self.apply_rule(rule)),
+        }
     }
-}
 
-impl HttpContext for Filter {
-    fn on_http_request_headers(&mut self, _: usize, _: bool) -> Action {
-        let context_id = self.context_id();
-        info!("context #{}: on_http_request_headers called", context_id);
+    fn apply_rule(&self, rule: &Rule) -> bool {
+        if let Some(paths) = &rule.paths {
```

Comment:
So that was some of my reservation around the `Option` here… why not have an empty vec when nothing is there and have the field be `paths: Vec<String>`? Tho maybe it make the whole check slightly more awkward… arguably, now `Some(empty vec) != None`…

Reply:
I need to play a little bit more with serde, but I guess `Option` is needed to differentiate between:
* `rules` not defined
```yaml
gateway_actions:
- configurations: {}
```
* rules defined but empty
 ```yaml
gateway_actions:
- rules: []
  configurations: {}
```

Reply:
Are those two equivalent tho? They are right?

Reply:
For the kuadrant's purposes, they are equivalent. I have not checked (yet) if the de-serialization works the same if Option is removed. My guess is that it does not work the same.

Reply:
Well. that's up to us. 

if I define
```yaml
- rules: []
```
What is expected? the spec says at least one set of rules must match. 

```
Matching occurs when at least one rule applies against the incoming request. If rules are not set, it is equivalent to matching all the requests.
```
Maybe the current implementation is right. If no rules are set, then the rate limit config is applied. If empty rules is set, then none applies, therefore the rate limit configuration should not apply.

wdyt?

Reply:
Hmm my initial instinct when reading this is that ``` - rules: [] ``` is equiv to not specifying rules. At least that is the way I read those two examples.

Reply:
for `rules`, `[]` == `nil` meaning that rate limit configurations will apply unconditionally. I think that `Option` will not be needed. 

Only when there is at least one rule, matching occurs when at least one rule applies

Reply:
implemented